### PR TITLE
Alternate ALARM PWM function

### PR DESCRIPTION
**Pull request by andyp1per**
Add support for allowing a different PWM channel to function as an alarm

### CODE REVIEW

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp [hunk 1]
```diff
@@ -77,6 +78,17 @@ void RCOutput::init()
             uint8_t chan = group.chan[j];
             if (SRV_Channels::is_GPIO(chan+chan_offset)) {
                 group.chan[j] = CHAN_DISABLED;
+            } else if (SRV_Channels::is_alarm(chan+chan_offset)) {
+                // alarm takes the whole timer
+                group.chan[j] = CHAN_DISABLED;
```
Comment (peterbarker) [inline libraries/AP_HAL_ChibiOS/RCOutput.cpp:83]:
```suggestion
```

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp [hunk 2]
```diff
@@ -77,6 +78,17 @@ void RCOutput::init()
...
+                group.ch_mask = 0;
+                for (uint8_t k = 0; k < 4; k++) {
+                    if (group.chan[k] != CHAN_DISABLED) {
+                        group.chan[k] = CHAN_DISABLED;
+                    }
```
Comment (peterbarker) [inline libraries/AP_HAL_ChibiOS/RCOutput.cpp:88]:
```suggestion
                    group.chan[k] = CHAN_DISABLED;
```

**Review by peterbarker — COMMENTED** [review]
Should this instead be something like the hwdef specifies a default servo output channel that the alarm is on, and servo channel output sets the default for that channel to be `k_alarm`?

Wouldn't that potentially allow alarm to be useful on any board?

**Comment by andyp1per** [comment]
> Should this instead be something like the hwdef specifies a default servo output channel that the alarm is on, and servo channel output sets the default for that channel to be `k_alarm`?
> 
> Wouldn't that potentially allow alarm to be useful on any board?

The issue I am trying to solve is where there is no spare timer for the alarm. If a user is willing to give up a PWM output for the alarm then they can use that - but it cannot be the default

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp (hunk 2 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/RCOutput.cpp]: Fixed thanks

**Comment by andyp1per** [comment]
Works on my KakuteH7Mini on M5 with an active buzzer

**Comment by andyp1per** [comment]
Costs 300 bytes on Matek F405, less on other boards

File: libraries/AP_HAL_ChibiOS/Util.cpp [hunk 3]
```diff
@@ -169,24 +171,39 @@ bool Util::toneAlarm_init(uint8_t types)
...
     return false;
 #else
     return true;
 #endif
 }
 
+#if HAL_USE_PWM == TRUE
+bool Util::toneAlarm_init(const PWMConfig& pwm_cfg, PWMDriver* pwm_drv, pwmchannel_t chan, bool active_high)
+{
+    _toneAlarm_pwm_group.pwm_cfg = pwm_cfg;
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/Util.cpp:185]: looks like we can end up with both running, need to shut down the existing pwm output

**Review by tridge — CHANGES_REQUESTED** [review]
need testing on a board with a buzzer, both using normal buzzer and assigned to a pin, so taking over

File: libraries/AP_HAL_ChibiOS/Util.cpp [hunk 4]
```diff
@@ -169,24 +171,39 @@ bool Util::toneAlarm_init(uint8_t types)
...
+    _toneAlarm_pwm_group.pwm_drv = pwm_drv;
+    _toneAlarm_pwm_group.pwm_cfg.period = 1000;
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/Util.cpp:187]: we might be able to use the existing clock, and scale the period, so we don't need to shutdown the other channels

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_HAL_ChibiOS/Util.cpp (hunk 3 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/Util.cpp:188]: Fixed

**Comment by andyp1per** [comment]
@tridge tested now. My testing showed up a pre-existing bug: e2b232f84ab9d1aa7bc1840a5d98c5312ebfa808 not sure how serious it is but it means GPIO channels are not properly disabled on FC with IOMCU currently.

**Comment by tridge** [comment]
@andyp1per we should discuss this more

**Review by tridge — APPROVED** [review]

**Pull request merged**